storage/memetcd/peer_store.go

```diff
@@ -0,0 +1,535 @@
+
+func (ps *peerStore) DeleteSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	if err := ps.memoryStore.DeleteSeeder(ih, p); err != nil {
+		return err
+	}
+
+	if ps.cfg.EtcdOpAsync {
+		go func() {
+			if err := ps.etcdDeletePeer(ih, p, true); err != nil {
+				log.Error(err)
+			}
+		}()
+	} else {
+		if err := ps.etcdDeletePeer(ih, p, true); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ps *peerStore) PutLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	if err := ps.memoryStore.PutLeecher(ih, p); err != nil {
+		return err
+	}
+
+	if ps.cfg.EtcdOpAsync {
+		go func() {
+			if err := ps.etcdPutOrUpdatePeer(ih, p, false); err != nil {
+				log.Error(err)
+			}
+		}()
+	} else {
+		if err := ps.etcdPutOrUpdatePeer(ih, p, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ps *peerStore) DeleteLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	if err := ps.memoryStore.DeleteLeecher(ih, p); err != nil {
+		return err
+	}
+
+	if ps.cfg.EtcdOpAsync {
+		go func() {
+			if err := ps.etcdDeletePeer(ih, p, false); err != nil {
+				log.Error(err)
+			}
+		}()
+	} else {
+		if err := ps.etcdDeletePeer(ih, p, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ps *peerStore) GraduateLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	if err := ps.memoryStore.GraduateLeecher(ih, p); err != nil {
+		return err
+	}
+
+	if ps.cfg.EtcdOpAsync {
+		go func() {
+			if err := ps.etcdGraduateLeecher(ih, p); err != nil {
+				log.Error(err)
+			}
+		}()
+	} else {
+		if err := ps.etcdGraduateLeecher(ih, p); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ps *peerStore) AnnouncePeers(ih bittorrent.InfoHash, seeder bool, numWant int, announcer bittorrent.Peer) (peers []bittorrent.Peer, err error) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	peers, err = ps.memoryStore.AnnouncePeers(ih, seeder, numWant, announcer)
+	return
+}
+
+func (ps *peerStore) ScrapeSwarm(ih bittorrent.InfoHash, addressFamily bittorrent.AddressFamily) (resp bittorrent.Scrape) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	resp = ps.memoryStore.ScrapeSwarm(ih, addressFamily)
+	return
+}
+
+// We do serializable reads of announce keys (compared to linearizable by default) for lower latency
+// Obvious con is that stale reads might occur which we are fine with in this scenario
+func (ps* peerStore) Bootstrap(...interface{}) error {
```

Comment:
Yes there is. because I am not sure what bootstrap can be used else for. There could be some implementation which bootstraps some scrape statistics or some other data which was required to be persisted across restarts.

Another approach is to explicitly call this `BootstrapAnnounces`. Does this make more sense or my former argument?

Reply:
in fact announce is also a struct here specific to memory store. what if some other store persists announces differently?

Reply:
ignore, got rid of bootstrap